### PR TITLE
Domain Selector Reviewed

### PR DESCRIPTION
**Pull request by qsm-odoo**

### CODE REVIEW

File: addons/web/static/src/js/widgets/domain_selector.js [hunk 1]
```diff
@@ -0,0 +1,558 @@
...
+var pyeval = require("web.pyeval");
+var session = require("web.session");
+var Widget = require("web.Widget");
+
+var _t = core._t;
+
+// "child_of", "parent_of", "like", "not like", "=like", "=ilike"
+// are only used if user entered them manually or if got from demo data
+var operator_mapping = {
+    "=": _t("is equal to"),
```
Comment (ged-odoo) [inline addons/web/static/src/js/widgets/domain_selector.js]: _lt

File: addons/web/static/src/js/widgets/domain_selector.js [hunk 2]
```diff
@@ -0,0 +1,558 @@
...
+    /// the widget is currently representing (an array).
+    getDomain: function () {},
+});
+/// The DomainTree is a DomainNode which can handle subdomains (a domain which is composed
+/// of multiple parts). It thus will be composed of other DomainTree instances and/or leaf parts
+/// of a domain (@see DomainLeaf).
+var DomainTree = DomainNode.extend({
+    template: "DomainTree",
+    events: _.extend({}, DomainNode.prototype.events, {
+        "click .o_domain_tree_operator_selector > ul > li > a": function (e) {
```
Comment (ged-odoo) [inline addons/web/static/src/js/widgets/domain_selector.js]: humm, really specific here...

File: addons/web/static/src/js/widgets/domain_selector.js [hunk 3]
```diff
@@ -0,0 +1,558 @@
...
+        var start = 1;
+        if (!_.contains(["&", "|", "!"], this.operator)) {
+            this.operator = "&";
+            start = 0;
+        }
+
+        // Add flattened children by search the appropriate number of children in the rest
+        // of the domain (after the operator)
+        var childrenDomains = domain.slice(start);
+        var toFound = 1;
```
Comment (ged-odoo) [inline addons/web/static/src/js/widgets/domain_selector.js]: beautiful english here...

File: addons/web/static/src/js/widgets/domain_selector.js [hunk 4]
```diff
@@ -0,0 +1,558 @@
...
+    /// @param afterNode - the node after which the new domain part must be added (at the end if not given)
+    /// @trigger_up domain_changed if the child is added
+    /// @return true if the part was added, false otherwise (the afterNode was not found)
+    addChild: function (domain, afterNode) {
+        var i = afterNode ? _.indexOf(this.children, afterNode) : this.children.length;
+        if (i < 0) return false;
+
+        this.children.splice(i+1, 0, instantiateNode(this, this.model, domain, this.options));
+        this.trigger_up("domain_changed", {child: this});
+        return true;
```
Comment (ged-odoo) [inline addons/web/static/src/js/widgets/domain_selector.js]: not impressed by a method that always return true

File: addons/web/static/src/js/widgets/domain_selector.js [hunk 5]
```diff
@@ -0,0 +1,558 @@
...
+    /// @trigger_up domain_changed if the child is removed
+    /// @return true if the child was removed, false otherwise (the widget does not own the child)
+    removeChild: function (oldChild) {
+        var i = _.indexOf(this.children, oldChild);
+        if (i < 0) return false;
+
+        this.children[i].destroy();
+        this.children.splice(i, 1);
+        this.trigger_up("domain_changed", {child: this});
+        return true;
```
Comment (ged-odoo) [inline addons/web/static/src/js/widgets/domain_selector.js]: idem

File: addons/web/static/src/js/widgets/model_field_selector.js [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+        this.model = model;
+        this.chain = chain;
+        this.options = {filters: filters || {}};
+
+        this.pages = [];
+        this.selectedField = false;
+        this.isSelected = true;
+        this.dirty = false;
+
+        this.debug = session.debug;
```
Comment (ged-odoo) [inline addons/web/static/src/js/widgets/model_field_selector.js]: as much as possible, widgets should be instantiated with all the necessary information, not fetch the information themselves.

File: addons/web/static/src/js/views/form_widgets.js [hunk 7]
```diff
@@ -383,87 +384,111 @@ var FieldFloat = FieldChar.extend({
...
+            this.set_value(this.domainSelector.getDomain(), true);
+        },
+    },
+    init: function () {
         this._super.apply(this, arguments);
+
+        this.valid = true;
         this.debug = session.debug;
+        this.options = _.defaults(this.options || {}, {
+            model: "res.partner",
```
Comment (qsm-odoo) [inline addons/web/static/src/js/views/form_widgets.js]: -> undefined

File: addons/web/static/src/js/views/form_widgets.js [hunk 8]
```diff
@@ -383,87 +384,111 @@ var FieldFloat = FieldChar.extend({
...
+            fs_filters: {},
```
Comment (qsm-odoo) [inline addons/web/static/src/js/views/form_widgets.js]: :+1:

File: addons/web/static/src/js/views/form_widgets.js [hunk 9]
```diff
@@ -383,87 +384,111 @@ var FieldFloat = FieldChar.extend({
...
-        if (this.options.model_field){
-            this.field_manager.fields[this.options.model_field].on("change:value", this, function(){
-                if (self.view && self.view.record_loaded.state == "resolved" && self.view.onchanges_mutex){
-                    self.view.onchanges_mutex.def.then(function(){
-                        self.render_value();
-                    });
-                }
-            });
+        this.model = _get_model.call(this); // TODO get the model another way ?
+        this.field_manager.on("view_content_has_changed", this, function () {
```
Comment (qsm-odoo) [inline addons/web/static/src/js/views/form_widgets.js]: Yep

File: addons/web/static/src/js/views/form_widgets.js [hunk 10]
```diff
@@ -383,87 +384,111 @@ var FieldFloat = FieldChar.extend({
...
-            on_selected: function(selected_ids) {
-                if (!self.get('effective_readonly')) {
-                    self.set_value(dialog.get_domain(selected_ids));
-                }
-            }
+        // Show number of selected records
+        new Model(this.model).call("search_count", [domain], {
+            context: this.build_context(),
+        }).then((function (data) {
+            this.valid = true;
```
Comment (qsm-odoo) [inline addons/web/static/src/js/views/form_widgets.js]: Don't like it either but today, I prefer bind(this) to self (we will see tomorrow if that changes :))

File: addons/web/static/src/js/views/form_widgets.js [hunk 11]
```diff
@@ -383,87 +384,111 @@ var FieldFloat = FieldChar.extend({
...
+        }).bind(this), (function (error, e) {
+            e.preventDefault();
+            this.valid = false;
+        }).bind(this)).always((function (data) {
+            this.$recordsCountDisplay.text(data || 0);
+            this.$showSelectionButton.toggleClass("hidden", !this.valid);
+            this.$errorMessage.toggleClass("hidden", this.valid);
+        }).bind(this));
+    },
+    is_syntax_valid: function() {
```
Comment (qsm-odoo) [inline addons/web/static/src/js/views/form_widgets.js]: Will be removed in the new Field API

File: addons/web/static/src/js/widgets/domain_selector.js [hunk 12]
```diff
@@ -0,0 +1,558 @@
+odoo.define("web.DomainSelector", function (require) {
+"use strict";
+
+var core = require("web.core");
+var datepicker = require("web.datepicker");
+var formats = require ("web.formats");
+var ModelFieldSelector = require("web.ModelFieldSelector");
+var pyeval = require("web.pyeval");
+var session = require("web.session");
```
Comment (qsm-odoo) [inline addons/web/static/src/js/widgets/domain_selector.js]: :+1:

File: addons/web/static/src/js/widgets/domain_selector.js (hunk 1 above)
Comment (qsm-odoo) [inline addons/web/static/src/js/widgets/domain_selector.js]: :+1:

File: addons/web/static/src/js/widgets/domain_selector.js (hunk 2 above)
Comment (qsm-odoo) [inline addons/web/static/src/js/widgets/domain_selector.js]: Seems OK to me ("click on links in the dropdown of the selection group")

File: addons/web/static/src/js/widgets/domain_selector.js (hunk 3 above)
Comment (qsm-odoo) [inline addons/web/static/src/js/widgets/domain_selector.js]: ... oups :D :+1:

File: addons/web/static/src/js/widgets/domain_selector.js (hunk 4 above)
Comment (qsm-odoo) [inline addons/web/static/src/js/widgets/domain_selector.js]: I think you did not look at it right, there is a `return false` above

File: addons/web/static/src/js/widgets/domain_selector.js (hunk 5 above)
Comment (qsm-odoo) [inline addons/web/static/src/js/widgets/domain_selector.js]: idem

File: addons/web/static/src/js/widgets/model_field_selector.js (hunk 6 above)
Comment (qsm-odoo) [inline addons/web/static/src/js/widgets/model_field_selector.js]: :+1:

**Comment by qsm-odoo** [comment]
Merged with commits:
community: https://github.com/odoo/odoo/commit/e3f99885ca789b338a4cb9ab9a154db50dff500d
enterprise: https://github.com/odoo/enterprise/commit/fc179b7d172822b76488eaba4776c0a99a417b6b

The temporary "[FIX] gamification" commit still has to be merged (but it is maybe for 9.0/10.0, I must check).

**Pull request closed without merging**